File: android/src/main/java/com/oj/plugins/offlinebundle/OfflineBundlePlugin.kt
package com.oj.plugins.offlinebundle

import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "OfflineBundle")
class OfflineBundlePlugin : Plugin() {
    private val tag = "OfflineBundlePlugin"
    private var manager: OfflineBundleManager? = null

    override fun load() {
        try {
            Log.i(tag, "load() invoked")
            manager = OfflineBundleManager(
                context = context,
                bridge = bridge,
            )
            runCatching { manager?.loadLocalIfExists() }
                .onFailure { error ->
                    Log.e(tag, "loadLocalIfExists() failed during plugin load", error)
                }
            Log.i(tag, "load() completed")
        } catch (error: Throwable) {
            Log.e(tag, "load() fatal error", error)
            manager = OfflineBundleManager(
                context = context,
                bridge = bridge,
            )
        }
    }

    @PluginMethod
    fun getLocalBundleInfo(call: PluginCall) {
        call.resolve(manager?.getLocalBundleInfo()?.toJsObject() ?: JSObject())
    }

    @PluginMethod
    fun getBuiltinBundleInfo(call: PluginCall) {
        call.resolve(manager?.getBuiltinBundleInfo()?.toJsObject() ?: JSObject())
    }

    @PluginMethod
    fun getBundleDownloadPath(call: PluginCall) {
        val fileName = call.getString("fileName").orEmpty()
        val path = manager?.getBundleDownloadPath(fileName).orEmpty()
        call.resolve(JSObject().put("path", path))
    }

    @PluginMethod
    fun installBundle(call: PluginCall) {
        val zipPath = call.getString("zipPath").orEmpty()
        val installed = manager?.installBundle(zipPath) ?: false
        call.resolve(JSObject().put("installed", installed))
    }

    @PluginMethod
    fun applyUpdate(call: PluginCall) {
        val applied = manager?.applyUpdateIfReady() ?: false
        call.resolve(JSObject().put("applied", applied))
    }
}

File: ios/Sources/OfflineBundlePlugin/OfflineBundlePlugin.swift
import Capacitor

@objc(OfflineBundlePlugin)
public class OfflineBundlePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OfflineBundlePlugin"
    public let jsName = "OfflineBundle"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getLocalBundleInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBuiltinBundleInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBundleDownloadPath", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "installBundle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "applyUpdate", returnType: CAPPluginReturnPromise)
    ]

    private var manager: OfflineBundleManager?

    public override func load() {
        manager = OfflineBundleManager(
            viewController: bridge?.viewController as? CAPBridgeViewController
        )
        manager?.loadLocalIfExists()
        NSLog("[OfflineBundlePlugin] load completed")
    }

    @objc func getLocalBundleInfo(_ call: CAPPluginCall) {
        call.resolve(manager?.getLocalBundleInfo().toDictionary() ?? [:])
    }

    @objc func getBuiltinBundleInfo(_ call: CAPPluginCall) {
        call.resolve(manager?.getBuiltinBundleInfo().toDictionary() ?? [:])
    }

    @objc func getBundleDownloadPath(_ call: CAPPluginCall) {
        let fileName = call.getString("fileName") ?? ""
        let path = manager?.getBundleDownloadPath(fileName: fileName) ?? ""
        call.resolve(["path": path])
    }

    @objc func installBundle(_ call: CAPPluginCall) {
        let zipPath = call.getString("zipPath") ?? ""
        let installed = manager?.installBundle(zipPath: zipPath) ?? false
        call.resolve(["installed": installed])
    }

    @objc func applyUpdate(_ call: CAPPluginCall) {
        let applied = manager?.applyUpdateIfReady() ?? false
        call.resolve(["applied": applied])
    }
}
